<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Charting human biology using AI for Precision Health &amp; Precision Medicine - Workshop: Charting human biology using AI for Precision Health &amp; Precision Medicine</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./index.html">Home</a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-center sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Charting human biology using AI for Precision Health &amp; Precision Medicine</a> 
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">Home</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./schedule.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Schedule</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./materials.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Call for poster presentations</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./organizers.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Organizers</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#location" id="toc-location" class="nav-link active" data-scroll-target="#location">Location</a></li>
  <li><a href="#time" id="toc-time" class="nav-link" data-scroll-target="#time">Time</a></li>
  <li><a href="#what-is-it-about" id="toc-what-is-it-about" class="nav-link" data-scroll-target="#what-is-it-about">What is it about?</a></li>
  <li><a href="#confirmed-speakers" id="toc-confirmed-speakers" class="nav-link" data-scroll-target="#confirmed-speakers">Confirmed speakers</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Workshop: Charting human biology using AI for Precision Health &amp; Precision Medicine</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<p><img src="images/charting_human_biology_using_AI.png" class="img-fluid" style="width:35.0%"></p>
<section id="location" class="level3">
<h3 class="anchored" data-anchor-id="location">Location</h3>
<p>238 Main Street, Cambridge, MA, 02139</p>
<p>6th floor conference space where <a href="https://www.bayer.com/en/">Bayer</a>’s office is located.</p>
</section>
<section id="time" class="level3">
<h3 class="anchored" data-anchor-id="time">Time</h3>
<p>Nov 16th 2023 , 8:30am - 4:15pm <a href="./schedule.html">schedule</a> - <a href="./materials.html">call for posters / registration</a></p>
</section>
<section id="what-is-it-about" class="level3">
<h3 class="anchored" data-anchor-id="what-is-it-about">What is it about?</h3>
<p>The human biology is arguably one of the most precious complex systems in existence that is both multi-scale and highly dynamic in nature. Though a full understanding of such a system eludes us at the present, modern medicine relies on sophisticated biosensors and data analytics methods. Generation of actionable insights into health and disease requires an interdisciplinary effort bringing together domain scientists and technologists.</p>
<p>With the unprecedented growth of available molecular assays (e.g.&nbsp;single cell m’omics, spatial omics, proteogenomics, and so on) and increasing availability of real world data, there is an essential need to develop methods and expertise in multi-omics analysis and integration, specially using ML/AI to bring together prior knowledge, disease biology and clinical data to actionable insights for drug discovery, diagnostics and prediction of response and resistance to treatments.</p>
<p>In this mini-conference, we will focus on multi-omics technologies and corresponding novel analytics approaches to help build integrative models of the systems biology in order to understand the healthy state and transition into a disease state, and how that differs between subgroups of individuals leading to development of more effective interventions specific to each subgroup.</p>
</section>
<section id="confirmed-speakers" class="level3">
<h3 class="anchored" data-anchor-id="confirmed-speakers">Confirmed speakers</h3>
<p><strong>Session 1</strong></p>
<p>Virginia Savova, PhD. Senior Director, Global Head of Single-cell Biology</p>
<p><strong>Talk Title:</strong></p>
<p><strong>From Single-cell Biology To Cell-Centric Drug Discovery</strong></p>
<p><strong>Abstract</strong></p>
<p>TBD</p>
<hr>
<p><img src="images/matthew.png" class="img-fluid" style="width:30.0%"></p>
<p>Dr.&nbsp;Matthew Bernstein is a Principal Scientist, Computational Biology at Immunitas Therapeutics. He develops computational and statistical methods to analyze large scale single-cell and spatial transcriptomics datasets to discover new immunotherapies for challenging cancers. Prior to joining Immunitas, Dr.&nbsp;Bernstein completed a Ph.D.&nbsp;in Computer Sciences from the University of Wisconsin – Madison and a postdoctoral fellowship at the Morgridge Institute for Research.</p>
<p><strong>Talk Title:</strong></p>
<p><strong>Annotating cell types and cell states in single-cell omics data: a challenging problem</strong></p>
<p><strong>Abstract:</strong></p>
<p>A first and fundamental task in the analysis of single-cell omics data is annotating the cell type or phenotype of each cell. Though fundamental, this task still poses a bottleneck in the analysis of these data despite the fact that a large number of machine learning based methods have been developed to facilitate this task. Part of the reason for these difficulties is the lack of a standard definition for what constitutes a cell type versus a cell state and a lack of standard nomenclature for cataloging them. In this talk, I will present a mental framework for thinking about cell types versus cell states, how we can catalogue them using ontologies, and how various existing machine learning methods implicitly define the cell type annotation task.</p>
<hr>
<p><img src="images/Shantanu.jpg" class="img-fluid" style="width:30.0%"></p>
<p><a href="https://www.linkedin.com/in/shntnu/">Shantanu Singh, Ph.D.</a>, is a Senior Group Leader at the Broad Institute, where he serves as the co-head of the <a href="https://carpenter-singh-lab.broadinstitute.org/">Carpenter–Singh lab</a>. His lab is dedicated to discovering new disease therapies using cell imaging to advance our understanding of biological processes. He is passionate about bridging the gaps between rapidly developing fields in high-dimensional biology to address the significant challenges in improving human health.</p>
<p><strong>Talk Title:</strong></p>
<p><strong>Finding disease phenotypes and candidate therapeutics using images: Cell Painting</strong></p>
<p><strong>Abstract:</strong></p>
<p>Cell morphology holds valuable information about the mechanisms and functions of cell structures, and microscopy has been instrumental in uncovering biological phenomena. However, images of cell structures contain more information than what meets the eye. With the advancements in image analysis and deep learning, this rich information can be captured and quantified to drive progress in basic biology research and drug discovery. Image-based profiling using fluorescence microscopy assays such as Cell Painting can reveal the impact of diseases, drugs, and genes on cells, uncover mechanisms of action for small molecules, identify disease-associated phenotypes, detect response to drugs, and predict biological impact and toxicity of compounds, among other applications. This is leading to a growing impact on the pharmaceutical industry as cell morphology becomes a powerful data source for systems biology alongside molecular omics readouts.</p>
<hr>
<p><strong>Session 2</strong></p>
<p><img src="images/Elisabeth_Wood.png" class="img-fluid" style="width:30.0%"></p>
<p>Elizabeth Wood, PhD, currently serves as President and Chief Executive Officer of JURA Bio, Inc.&nbsp;and serves on the company’s Board of Directors. JURA is transforming the AI-driven design of therapeutic candidates across multiple modalities like proteins, cells and gene therapies by extending machine learning from upstream generative gene synthesis to downstream clinical development. Dr.&nbsp;Wood serves on the steering committee of the Broad Institute’s Models, Inference &amp; Algorithms (MIA) initiative. From 2019-2022, she led LMRL, a NeurIPS workshop on machine learning applications in biology. She serves on the Board of Project Clio, a 501(c)3 dedicated to furthering research of autoimmune diseases in underserved populations, as well as on the board of the MIT Safe Household Water Project. She is a 2022 Henri Termeer Fellow. Dr.&nbsp;Wood was previously a postdoctoral fellow at Harvard University working with Dr.&nbsp;Adam Cohen, as well as a visiting researcher with Kresten Lindorff-Larsen at the Linderstrøm-Lang Centre for Protein Science at the University of Copenhagen. She received her PhD in physics from the Danish Technical University and AB in physics from Harvard College.</p>
<p><strong>Talk Title:</strong></p>
<p><strong>Variational Synthesis Atlas of Human T Cell Repertoires (VISTA-T): Generative Library Design Enables Large Scale Discovery of Therapeutically Relevant TCRs</strong></p>
<p><strong>Abstract:</strong></p>
<p>A major challenge in adoptive T cell immunotherapy is the identification of natural T cell receptors (TCRs) with both high activity and tumor antigen specificity. Synthetic TCRs, despite their promise, are fraught with risks such as potential cross-reactivity and the unpredictability of the relationship between TCR binding affinity and activity in response to antigen (peptide-MHC). To address this challenge, we introduce Jura’s VISTA platform, a method for therapeutic discovery that merges computational design, genome editing, and single-cell sequencing. The platform replicates the natural population-wide diversity of human TCRs and assesses their binding activity at a high-throughput scale. To build TCR libraries, VISTA-T implements variational synthesis, a recently proposed method for efficiently synthesizing samples from a generative sequence model in the laboratory. Here, the generative models are trained on a wide array of human TCR repertoires. The resulting libraries consist of highly realistic TCRs, at a scale and diversity comparable to natural repertoires (&gt;10^9 unique sequences). These TCRs are assessed via a high-throughput binding assay in a human T cell line, with readout of downstream activation. VISTA-T successfully identifies TCRs specific to various clinically relevant tumor-associated antigens, validating their translational potential.</p>
<hr>
<p><img src="images/xihao.png" class="img-fluid" style="width:30.0%"></p>
<p><a href="https://www.linkedin.com/in/xihao-sherlock-hu/">Dr.&nbsp;Sherlock Hu</a>, the CIO of <a href="https://gv20oncotherapy.com/">GV20 Therapeutics</a>, earned his Ph.D.&nbsp;in Computer Science from the Chinese University of Hong Kong. He later pursued postdoctoral training in Prof.&nbsp;Shirley Liu’s lab at Dana-Farber Cancer Institute and Harvard University. In 2019, he was named Parker Scholar by the Parker Institute for his innovative work in mining tumor-infiltrating B cell receptor repertoires. Dr.&nbsp;Hu’s expertise in Computational Biology and Immunology underpins GV20’s AI platform, enabling the deep learning-based de novo design of fully human functional antibodies. At GV20, Sherlock leads the bioinformatics and AI teams, spearheading AI-driven antibody drug design.</p>
<p><strong>Talk Title:</strong></p>
<p><strong>AI for Target Identification and Antibody Discovery</strong></p>
<p><strong>Abstract:</strong></p>
<p>In cancer research, the exploration of innovative strategies for identifying therapeutic targets and developing effective treatments remains paramount. This presentation delves into a novel approach that harnesses the power of artificial intelligence (AI) to revolutionize cancer therapeutics. By modeling the intricate landscape of antibody-antigen interactions within the tumor microenvironment, our research strives to uncover new treatments derived from patients’ immune systems.</p>
<p>Central to our methodology is the integration of AI, which enables the rational discovery of cancer targets and functional antibodies. As an example, we present a first-in-class antibody drug that exhibits an excellent pre-clinical profile and is currently advancing through clinical trials. Our findings not only shed light on the intricate web of interactions within the tumor microenvironment but also underscore the viability of AI-driven strategies in addressing challenging issues in precision medicine.)</p>
<hr>
<p><img src="images/Michal_Warchol.jpg" class="img-fluid" style="width:30.0%"></p>
<p>Dr.&nbsp;Michal Warchol, is the Executive Vice President at Ardigen. He serves as the General Director of AI Lab that specializes in providing AI-driven technology solutions to biotech and pharmaceutical companies, accelerating the drug discovery processes by transforming large volumes of data into actionable scientific insights. Dr.&nbsp;Warchoł did a PhD in Statistics at the Institute of Statistics, Biostatistics and Actuarial Sciences at Université Catholique de Louvain in Belgium. He conducted part of his research program at the Department of Statistics at Columbia University in New York.</p>
<p><strong>Talk title:</strong></p>
<p><strong>Multimodal approach to phenotype-guided virtual screening of small molecules</strong></p>
<p><strong>Abstract:</strong> High Content Screening (HCS) is a well-established technology used in the drug discovery process. Recent advancements in Artificial Intelligence, computer vision, and computational capabilities increase phenotypic screening potential; leveraging the massive amounts of information encoded in multicolor images at single-cell resolution. Our research is centered on the synergistic integration of chemical structures and images from HCS experiments, resulting in a more robust data structure. We demonstrate a substantial performance enhancement achieved through the multimodal approach in diverse applications, including chemical and biological properties prediction, hit identification and virtual screening of molecules that are most likely to induce a phenotype of interest.</p>
<hr>
<p><strong>Session 3</strong></p>
<p><img src="images/Gustavo.png" class="img-fluid" style="width:30.0%"></p>
<p>Gustavo Arango-Argoty, Ph.D.&nbsp;is an associate director at the Applied Data Science (ADS) team in the oncology data science department at AstraZeneca leading the deep learning research line withing the ADS group. His focus lies in developing explainable AI models for biomarker discovery. In particular, applying attention-based neural networks for biomarker identification, self-supervised learning, and transfer learning to gain insights from large real world evidence data to inform small size datasets including clinical trials.</p>
<p><strong>Talk title:</strong></p>
<p><strong>Pretrained transformers applied to clinical studies improve predictions of treatment efficacy and associated biomarkers</strong></p>
<p><strong>Abstract:</strong> Cancer treatment has made significant advancements in recent decades, leading to improved outcomes and quality of life for many patients. Despite the array of available therapies, including targeted, hormone, and checkpoint blockade immunotherapy, many patients experience treatment failure or eventual resistance. Attempts to predict the efficacy of therapies, particularly immuno-oncology therapies, have suffered from limited accuracy and difficulties in identifying molecular and other determinants of response. Improving treatment prediction alone is insufficient to create clinically meaningful research tools; additional prerequisites for this goal involve accommodating small data sets, effectively handling sparse features, integrating diverse clinical data, addressing missing measurements, ensuring interpretability, and extracting valuable biological insights for both clinical context and further research. Multimodal deep-learning models offer a promising avenue to surmount these challenges by leveraging their capacity and flexibility to learn from expansive and varied clinical and molecular data sets. Similar to their application in natural language and other domains, deep-learning models can uncover complex relationships within data that are pertinent to survival and treatment response. In this study, we introduce an explainable transformer-based deep-learning framework that addresses these challenges. This framework yields predictions of survival outcomes, as quantified by concordance index, that surpass the performance of state-of-the-art methods such as Cox proportional hazards, survival random forest, and tumor mutation burden, across diverse independent data sets. We developed the clinical transformer, a deep neural-network survival prediction framework that capitalizes on the flexibility of the deep-learning model, including training strategies like gradual and transfer learning, to maximize the use of available data to enhance survival predictions and generate actionable biological insights. Finally, we illustrate the future potential of the clinical transformer’s generative capability in early-stage clinical studies. By perturbing molecular features associated with immune checkpoint inhibition treatment in immunotherapy-naive patient profiles, we identified a subset of patients who may benefit from immunotherapy. These findings were subsequently validated across three independent immunotherapy treatment cohorts. We anticipate that this research will empower the scientific community to further harness data for the benefit of patients.</p>
<hr>
<p><img src="images/james.jpg" class="img-fluid" style="width:30.0%"></p>
<p><a href="https://www.linkedin.com/in/jamesrlindsay/">Dr.James Lindsay</a> is the Director of Software Engineering in the Department of Data Science at the Dana-Farber Cancer Institute. He helps build software to facilitate precision cancer medicine such as automated clinical trial matching, clinical DNA sequencing and next generation digital pathology imaging. Additionally he helps build software for cancer genomics research including the cBioPortal.</p>
<p><strong>Talk title:</strong></p>
<p><strong>Leveraging artificial intelligence predictions of treatment change to time out reach for enrollment in precision medicine clinical trials</strong></p>
<p><strong>Abstract:</strong></p>
<p>MatchMiner is an open source platform for cancer precision medicine, supporting clinical trial enrollments at Dana-Farber Cancer Institute. Clinical trials in precision oncology frequently face challenges with patient enrollment, partly due to the difficulty of finding patients which simultaneously meet eligibility criteria and are ready to change treatment. To address these challenges we initiated a study using AI tools to assist in identifying the right time to alert a physician about patient eligibility.</p>
<hr>
<p><img src="images/charles-fisher.jpg" class="img-fluid" style="width:30.0%"></p>
<p>Dr.&nbsp;Charles Fisher is a founder and CEO at Unlearn.AI. Unlearn develops generative models to create digital twins of individual patients that forecast their future health outcomes, and leverages these AI models to accelerate clinical research. Prior to Unlearn, Dr.&nbsp;Fisher was a theoretical physicist and software engineer working at the intersection of machine learning and biology. In industry, he worked as a machine learning engineer at a virtual reality startup and as a computational biologist at Pfizer. As an academic researcher, Dr.&nbsp;Fisher was a Phillippe Meyere Fellow in theoretical physics at École Normale Supérieure in Paris, France, and a postdoctoral scientist in theoretical biophysics at Boston University. He has a Ph.D.&nbsp;in biophysics from Harvard University and a B.S. in biophysics from the University of Michigan.</p>
<p><strong>Talk title:</strong></p>
<p><strong>Enabling Efficient Clinical Trials with Generative Models for Clinical Timeseries</strong></p>
<p><strong>Abstract:</strong> Clinical trials typically evaluate the efficacy of an investigational drug or device relative to placebo and/or standard of care. While randomization allows one to estimate the average difference between the treatment and control potential outcomes in the trial population, randomized studies require large numbers of participants to be assigned to a control group, which is both costly and inefficient and often against the wishes of patients themselves. Fortunately, the creation of large databases of individual participant data from historical studies and real world sources makes it possible to train generative models to forecast control potential outcomes for participants in clinical trials—which we call their ‘digital twins’. Participants’ digital twins can be used to create in silico control groups for single arm proof-of-concept studies, or to derive prognostic scores for use as ‘super covariates’ in randomized studies to increase power in ways that are suitable for phase 3 studies as illustrated by the EMA’s qualification opinion for Unlearn’s PROCOVA methodology.</p>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>